File: src/main/java/io/github/oleksivio/telegram/bot/api/model/method/location/SendLocation.java
package io.github.oleksivio.telegram.bot.api.model.method.location;

import com.fasterxml.jackson.annotation.JsonProperty;
import io.github.oleksivio.telegram.bot.api.model.NetworkError;
import io.github.oleksivio.telegram.bot.api.model.objects.std.keyboard.IKeyboard;
import io.github.oleksivio.telegram.bot.core.controller.network.ActionNetworker;
import io.github.oleksivio.telegram.bot.core.model.method.MessageAction;

/**
 * @see <a href="https://core.telegram.org/bots/api#sendlocation">sendLocation</a>
 */
public class SendLocation extends MessageAction {
    private static final String METHOD = "sendLocation";

    /**
     * latitude Float number Latitude of the location
     */
    @JsonProperty("latitude")
    private Double latitude = null;
    /**
     * longitude Float number Longitude of the location
     */
    @JsonProperty("longitude")
    private Double longitude = null;
    /**
     * live_period Integer Optional Period in seconds for which the location will be
     * updated (see Live Locations, should be between 60 and 86400
     */
    @JsonProperty("live_period")
    private Integer livePeriod = null;

    public SendLocation(ActionNetworker actionNetworker) {
        super(METHOD, actionNetworker);

    }

    public Double getLatitude() {
        return latitude;
    }

    public SendLocation setLatitude(Double latitude) {
        this.latitude = latitude;
        return this;
    }

    public Double getLongitude() {
        return longitude;
    }

    public SendLocation setLongitude(Double longitude) {
        this.longitude = longitude;
        return this;
    }

    public Integer getLivePeriod() {
        return livePeriod;
    }

    public SendLocation setLivePeriod(Integer livePeriod) {
        this.livePeriod = livePeriod;
        return this;
    }

    @Override
    public SendLocation setReplyMarkup(IKeyboard replyMarkup) {
        pSetReplyMarkup(replyMarkup);
        return this;
    }

    @Override
    public SendLocation setDisableNotification(Boolean disableNotification) {
        pSetDisableNotification(disableNotification);
        return this;
    }

    @Override
    public SendLocation setReplyToMessageId(Long replyToMessageId) {
        pSetReplyToMessageId(replyToMessageId);
        return this;
    }

    @Override
    public SendLocation setChatId(Long chatId) {
        pSetChatId(chatId);
        return this;
    }

    @Override
    public SendLocation setNetworkErrorListener(NetworkError onNetworkError) {
        pSetNetworkErrorListener(onNetworkError);
        return this;
    }
}

File: src/main/kotlin/io/github/oleksivio/telegram/bot/api/model/method/location/SendLocation.kt
package io.github.oleksivio.telegram.bot.api.model.method.location

import com.fasterxml.jackson.annotation.JsonProperty
import io.github.oleksivio.telegram.bot.api.model.objects.std.keyboard.IKeyboard
import io.github.oleksivio.telegram.bot.core.model.ApiDict
import io.github.oleksivio.telegram.bot.core.model.method.common.MessageAction

/**
 * @see [sendLocation](https://core.telegram.org/bots/api/#sendlocation)
 */
data class SendLocation(
        /**
         *  Unique identifier for the target chat or username of the target channel
         */
        @JsonProperty(ApiDict.CHAT_ID_KEY)
        override val chatId: Long,
        /**
         * latitude Float number Yes Latitude of new location
         */
        @JsonProperty(ApiDict.LATITUDE_KEY)
        val latitude: Double,
        /**
         * longitude Float number Yes Longitude of new location
         */
        @JsonProperty(ApiDict.LONGITUDE_KEY)
        val longitude: Double,
        /**
         * live_period Integer Optional Period in seconds for which the location will be
         * updated (see Live Locations, should be between 60 and 86400
         */
        @JsonProperty(ApiDict.LIVE_PERIOD_KEY)
        val livePeriod: Int? = null,
        /**
         * disable_notification Boolean Optional Sends the message silently.
         * Users will receive a notification with no sound.
         */
        @JsonProperty(ApiDict.DISABLE_NOTIFICATION_KEY)
        override val disableNotification: Boolean? = null,
        /**
         * reply_to_message_id Integer Optional If the message is a reply, ID of the original message
         */
        @JsonProperty(ApiDict.REPLY_TO_MESSAGE_ID_KEY)
        override val replyToMessageId: Long? = null,
        /**
         * reply_markup
         * InlineKeyboardMarkup or ReplyKeyboardMarkup or ReplyKeyboardRemove or ForceReply
         * Optional
         * Additional interface options.
         * A JSON-serialized object for an inline keyboard, custom reply keyboard,
         * instructions to remove reply keyboard or to force a reply from the user.
         */
        @JsonProperty(ApiDict.REPLY_MARKUP_KEY)
        override val replyMarkup: IKeyboard? = null

) : MessageAction() {
    @JsonProperty(ApiDict.METHOD_KEY)
    override val method = "sendLocation"

}
